timemachine/fe/single_topology.py

```diff
@@ -0,0 +1,783 @@
+    mol_a_proper_idxs = recursive_map(mol_a_pt.get_idxs(), a_to_c)
+    mol_a_improper_idxs = recursive_map(mol_a_it.get_idxs(), a_to_c)
+
+    all_dummy_bond_idxs = recursive_map(all_dummy_bond_idxs, b_to_c)
+    all_dummy_angle_idxs = recursive_map(all_dummy_angle_idxs, b_to_c)
+    all_dummy_proper_idxs = recursive_map(all_dummy_proper_idxs, b_to_c)
+    all_dummy_improper_idxs = recursive_map(all_dummy_improper_idxs, b_to_c)
+    all_dummy_x_angle_idxs = recursive_map(all_dummy_x_angle_idxs, b_to_c)
+    all_dummy_c_angle_idxs = recursive_map(all_dummy_c_angle_idxs, b_to_c)
+
+    # parameterize the combined molecule
+    mol_c_bond_idxs = mol_a_bond_idxs + all_dummy_bond_idxs
+    mol_c_bond_params = mol_a_bond_params + all_dummy_bond_params
+
+    mol_c_angle_idxs = mol_a_angle_idxs + all_dummy_angle_idxs
+    mol_c_angle_params = mol_a_angle_params + all_dummy_angle_params
+
+    mol_c_proper_idxs = mol_a_proper_idxs + all_dummy_proper_idxs
+    mol_c_proper_params = mol_a_proper_params + all_dummy_proper_params
+
+    mol_c_improper_idxs = mol_a_improper_idxs + all_dummy_improper_idxs
+    mol_c_improper_params = mol_a_improper_params + all_dummy_improper_params
+
+    mol_c_x_angle_idxs = all_dummy_x_angle_idxs
+    mol_c_x_angle_params = all_dummy_x_angle_params
+
+    mol_c_c_angle_idxs = all_dummy_c_angle_idxs
+    mol_c_c_angle_params = all_dummy_c_angle_params
+
+    return (
+        (mol_c_bond_idxs, mol_c_bond_params),
+        (mol_c_angle_idxs, mol_c_angle_params),
+        (mol_c_proper_idxs, mol_c_proper_params),
+        (mol_c_improper_idxs, mol_c_improper_params),
+        (mol_c_x_angle_idxs, mol_c_x_angle_params),
+        (mol_c_c_angle_idxs, mol_c_c_angle_params),
+    )
+
+
+class SingleTopologyV2:
+    def __init__(self, mol_a, mol_b, core, forcefield):
+        """
+        SingleTopology combines two molecules through a common core. The combined mol has
+        atom indices laid out such that mol_a is identically mapped to the combined mol indices.
+        The atoms in the mol_b's R-group is then glued on to resulting molecule.
+
+        Parameters
+        ----------
+        mol_a: ROMol
+            First ligand
+
+        mol_b: ROMol
+            Second ligand
+
+        core: np.array (C, 2)
+            Atom mapping from mol_a to to mol_b.
+
+        # ff: ff.Forcefield
+            # Forcefield to be used for parameterization.
+
+        """
+        assert mol_a is not None
+        assert mol_b is not None
+        self.mol_a = mol_a
+        self.mol_b = mol_b
+        self.core = core
+        self.ff = forcefield
+
+        assert core.shape[1] == 2
+
+        # map into idxs in the combined molecule
+        self.a_to_c = np.arange(mol_a.GetNumAtoms(), dtype=np.int32)  # identity
+        self.b_to_c = np.zeros(mol_b.GetNumAtoms(), dtype=np.int32) - 1
+
+        self.NC = mol_a.GetNumAtoms() + mol_b.GetNumAtoms() - len(core)
+
+        # mark membership in the combined molecule:
+        # 0: CORE
```

Comment:
Would it make sense to have an enum for this?

Reply:
sure - i don't see why not